### PR TITLE
Feature/insights stable version

### PR DESCRIPTION
**Pull request by q-litzler**
### Preliminary statements

As we want to release a stable version of Insights that does not rely on dependencies that are still in alpha (I'm looking at you, [WorkManager](https://developer.android.com/topic/libraries/architecture/workmanager/)), we want to use the [evernote](https://github.com/evernote/android-job) library to schedule jobs in the background.
Starting from `1.3.0`, the Evernote library will rely on the `WorkManager` itself, so we will be using evernote `1.2.6` in the meantime.

### Notes on JobScheduler

After investigating [JobScheduler](https://developer.android.com/reference/android/app/job/JobScheduler), I found that it was not suitable as it is only available above Android SDK 21, compared to Insights 14.

#### Periodic Job

The `Periodic` job will be in charge to upload events when these criteria are met:
- Maximum frequency of 15 minutes with 5 minutes oscillation
- The app is connected to the network

We want a simple and straightforward way to make sure there is only one periodic job scheduled at any point in time.

Each time Insights is created, we try the following logic:
- Check if there is a job ID stored in `SharedPreferences`
- If there is none, create the periodic job and store the job ID in it.
This has the following benefits:
- If the user deletes the app's data, both the the job ID and the job stored in the database will be deleted, which means both will be recreated next time `Insights` is instantiated.
- Only one periodic job should be stored and running at any time.

#### One time job

Each time an event is created, we attempt to upload it right away by creating a `OneTime` job.
If there is already a `OneTime` job running, we cancel it and replace it with a new one. This is done by using the `setUpdateCurrent(true)` method. We don't do anything if this attempt fail. The job is deletes itself from the database whether it was successful or not.

### Stetho

The demo app uses Stetho, which allows to take a peek inside the app's database to see exactly what is going on.
Launch your app on a device connected to your computer. Open your chrome browser and open this [url](chrome://inspect). The device and app should appear, open the inspection tool.
Go to the `Resource` tab, then open the `Web SQL` folder. The Evernote database should appear.

### CODE REVIEW

File: insights/src/main/kotlin/com/algolia/instantsearch/insights/event/EventUploaderEvernote.kt [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    private val keyUniquePeriodicJob = "PeriodicJob"
+
+    init {
+        JobManager.create(context).addJobCreator(EventJobCreator())
+    }
+
+    override fun startPeriodicUpload() {
+        val storedId = preferences.getInt(keyUniquePeriodicJob, -1)
+        if (storedId == -1) {
+            val repeatIntervalInMinutes = 15L
```
Comment (robertmogos) [inline insights/src/main/kotlin/com/algolia/instantsearch/insights/event/EventUploaderEvernote.kt]: Put the value as constants outside of the method

File: insights/src/main/kotlin/com/algolia/instantsearch/insights/event/EventUploaderEvernote.kt [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+            val flexTimeIntervalInMinutes = 5L
```
Comment (robertmogos) [inline insights/src/main/kotlin/com/algolia/instantsearch/insights/event/EventUploaderEvernote.kt]: same

File: insights/src/main/kotlin/com/algolia/instantsearch/insights/event/EventUploaderEvernote.kt [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+import com.evernote.android.job.JobManager
+import com.evernote.android.job.JobRequest
+import java.util.concurrent.TimeUnit
+
+
+internal class EventUploaderEvernote(context: Context) : EventUploader {
+
+    private val preferences = context.getSharedPreferences("Insights", Context.MODE_PRIVATE)
+
+    private enum class Preference {
```
[the remark this answers is not in the document]
Comment (q-litzler) [reply insights/src/main/kotlin/com/algolia/instantsearch/insights/event/EventUploaderEvernote.kt]: It is true than when you have multiple keys, enum are the way to go, as you can do some cool stuff with function extensions and enforcing compile safety with the close set defined by the enum. Which also mean I'd rather use enum everywhere for consistency, regardless if there is multiple keys or not.

**Comment by q-litzler** [comment]
@PLNech 
From reading the entire paragraph, I think you slightly misunderstood the line you mentionned.

> Depending on your requirements, this library decides which API suits your job. It provides a superset of all features from the existing AlarmManager, JobScheduler and GcmNetworkManager, e.g. you can run your job only if the device has a working Internet connection and is charging

It only says here than running a job only if the device has a working Internet connection and is charging is a feature, not the sole behaviour. The library offers an API for this:

```
 .setRequiresCharging(true)
 .setRequiresDeviceIdle(false)
.setRequiredNetworkType(JobRequest.NetworkType.CONNECTED)
```

:)

File: insights/src/main/kotlin/com/algolia/instantsearch/insights/event/EventUploaderEvernote.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (PLNech) [reply insights/src/main/kotlin/com/algolia/instantsearch/insights/event/EventUploaderEvernote.kt]: After our sync chat, agree that more typing/structure is welcome. Let's make sure thought o keep things as simple as possible, or to abstract the complexity via a minimal interface

**Pull request merged**